### ISSUE TITLE
Introduce global config for all widgets

### ISSUE THREAD

**Issue by pkozlowski-opensource** [ng-bootstrap/ng-bootstrap#518]
Good start described here: https://github.com/ng-bootstrap/ng-bootstrap/issues/510#issuecomment-236221937

We just need to refine it to have a consistent naming convention / way of coding it.

**Comment by jnizet** [comment]
We will also have to decide how these configuration services/classes are documented, and how the documentation of the components themselves must be written in order to identify which component inputs can be configured from the global configuration.

**Comment by pkozlowski-opensource** [comment]
So, finally had some more time to think about this item. Before diving into the proposal (which, in broad strokes, is identical to @jnizet one) here are the desired properties of the final solution:
* users should be able to configure (globally or in a given module / component tree) default values of inputs (those not provided when using a widget)
* we should still be able to properly generate API docs
* we shouldn't duplicate docs between config object and inputs 
* applying config in directives shouldn't result in code explosion (shouldn't be royal pain in the @#*&#)

Given the above and assuming that we want to have a global config for a directive named `NgbFoo` we should:
* have a class named `NgbFooConfig` that list all defaults for inputs (their names, types and default values)
* `NgbFoo` should have a constructor that optionally injects `NgbFooConfig` and propagates config values to input feilds

Example using `NgbProgressbar`: 

```typescript
export class NgbProgressbarConfig {
	animated = false;
	max = 100;
	striped = false
}
```

```typescript
export class NgbProgressbar {
    constructor(@Optional() config: NgbProgressbarConfig = new NgbProgressbarConfig()) {
        this.animated = config.animated; 
        ....  
    }
}
```

### Open questions
* I'm not sure where we should keep API docs... On inputs only? On config only? On both?
* ...

### Infrastructure changes
* change API generation script to extract info from classes
* change API generation script to merge info / docs from config classes and directives themselves
* decide how to present info about config in the API doc

**Comment by pkozlowski-opensource** [comment]
@jnizet I'm going to poke around infrastructure changes in the coming days but if you would like to send a PR introducing config for one of the widgets it would help us iterate on what needs to be done and answer all the remaining questions (timepicker is not the best candidate probably as there is quite a bit of changes in it atm)

**Comment by jnizet** [comment]
OK. I'll do that as quickly as possible (probably this evening, UTC+2). Just a note: in my attempt of doing that on the time picker, I first tried the syntax you suggest above:

```
    const DEFAULT_CONFIG = new NgbProgressbarConfig();
    export class NgbProgressbar {
        constructor(@Optional() config: NgbProgressbarConfig = DEFAULT_CONFIG) {
```

and Angular actually injected null (or undefined, I don't remember). I thus had to use 

```
    const DEFAULT_CONFIG = new NgbProgressbarConfig();
    export class NgbProgressbar {
        constructor(@Optional() config: NgbProgressbarConfig) {
            config = config || DEFAULT_CONFIG;
```

1. Is it a bug/limitation of Angular? 
2. Are you OK with that alternative solution?
3. Is it a good idea to use a unique default config object as I intended to do, since it's not supposed to be accessible from anywhere, instead of creating a new config object every time, or do you think it's a useless premature optimization?

**Comment by pkozlowski-opensource** [comment]
> OK. I'll do that as quickly as possible (probably this evening, UTC+2)

Awesome. But don't rush! Open source is more of a marathon than a sprint :-) 

>  Angular actually injected null (or undefined, I don't remember)

Oh right! Yeh, this is how DI operates, I don't think there is a way around this. So yes, your alternative is very valid.

> or do you think it's a useless premature optimization?

I would say that it is a premature optimization :-) Having a separate instance of a config object will free our minds from worrying about accidental mutations to this object. But in all honesty I don't have very strong opinion here.

**Comment by jnizet** [comment]
Now that RC5 and NgModule are a thing, and that there is a PR for documenting services in addition to components, can we move forward on this issue, which seems simple, yet important? 

I have already submitted a PR (#536), and am willing to continue with other components once we decide what to improve or change in the submitted code.

Besides, since modules can define their providers now and those providers are global, shouldn't we make each Config class a service and inject it in the components, instead of making it optional and creating an instance using `new` if not provided? That would simplify the component code, but most importantly, the application code wouldn't even have to add a provider to their own module in order to customize the configuration.

**Comment by pkozlowski-opensource** [comment]
@jnizet I think that we are still missing some changes to the API generation so that documentation is not duplicated in the directive and the config object.

> Besides, since modules can define their providers now and those providers are global, shouldn't we make each Config class a service and inject it in the components, instead of making it optional and creating an instance using new if not provided? 

Yes, I think we could try this now.

Honestly I'm knee-deep in modal / datepicker / rc6 so won't have time to help out with this particular issue for the next few days. If anyone wants to drive it it would be more than appreciated.

**Comment by jnizet** [comment]
I rebased the PR on master, and added another commit with the necessary infrastructure change. See the description of the PR #536 for details.

**Comment by icfantv** [comment]
I think we're just waiting on modal, popover, and typeahead at this point.

**Comment by jnizet** [comment]
I just submitted the PR for popover: #686.

dropdown doesn't have its config yet. `autoClose` should be in its config, but I'm not sure about `open`, because I can't see a use-case where someone would want dropdowns to be opened initially by default.

I don't think typeahead has any input that could have a default value. 

Modal could have default options, but these options are not inputs of a directive, so the pattern we're using for the components doesn't apply here. We could imagine changing the signature of `NgbModal.open()` by leaving the default value of options to undefined, and, if it's undefined, use an injected NgbModalConfig instance to create an instance of NgbModalOptions, but I'm not sure it's worth it. Developers could just define a service by themselves that creates their own default NgbModalOptions instance, and pass this to open(). 

@pkozlowski-opensource what do you think about these three points?

**Comment by icfantv** [comment]
I wasn't sure dropdown needed global config...but I suppose there's a use case for everything.

**Comment by pkozlowski-opensource** [comment]
@jnizet looks like you did dropdown part, here are my answers for the remaining 2 points.

> I don't think typeahead has any input that could have a default value.

I've landed support for `showHint` today, this is the only candidate I can think of.

> Modal could have default options, but these options are not inputs of a directive, so the pattern we're using for the components doesn't apply here. We could imagine changing the signature of NgbModal.open() by leaving the default value of options to undefined, and, if it's undefined, use an injected NgbModalConfig instance to create an instance of NgbModalOptions, but I'm not sure it's worth it. Developers could just define a service by themselves that creates their own default NgbModalOptions instance, and pass this to open().

Agree. Let's skip modal for now.

**Comment by pkozlowski-opensource** [comment]
@jnizet if I'm not mistaken the only eventual remaining part would by typeahead, right?

**Comment by jnizet** [comment]
@pkozlowski-opensource right. I plan to do it tonight, and that should flose this issue \o/

**Comment by piernik** [comment]
I'm wondering why not using global configuration in `forRoot` ?

**Comment by Halfist** [comment]
Is there any possibility to set `NgbModalOptions` globally so far, except for wrapping `open()` in a separate service?

**Comment by pkozlowski-opensource** [comment]
@Halfist nope, but we should add it. Could you please open a new issue / feature request for this? Thnx!

**Comment by Halfist** [comment]
@pkozlowski-opensource, sure, do you have some specific format for feature request issues? I saw few issues with "[FR]" and "(feature)" as a name prefix, but wanted to make it right and generic.

**Comment by pkozlowski-opensource** [comment]
@Halfist all is in the issue template (and no, we don't use any specific prefix / parts of title - it is free text)

**Comment by Halfist** [comment]
@pkozlowski-opensource, done. Please refer to #2487.

**Comment by maxokorokov** [comment]
@mtpultz, all configs will have `@Injecatble({providedin: 'root'})` when #2483 lands (most likely in 3.0.0)
